### ISSUE TITLE
`redundant-test-main-exit` assumes that all `os.Exit` calls are based on the return from `m.Run`

### ISSUE THREAD

**Issue by G-Rath** [mgechev/revive#1552]
**Describe the bug**
It seems that _any_ use of `os.Exit` in `TestMain` is flagged by `redundant-test-main-exit`, regardless of if the return from `m.Run` is actually use

**To Reproduce**
1. I updated revive `go install github.com/mgechev/revive@latest`
2. I run it with the following flags & configuration file:

```shell
# flags

revive .
```

```toml
# config file
[rule.redundant-test-main-exit]
```

```go
func TestMain(m *testing.M) {
	err := vcr.CleanCassettes()
	if err != nil {
		fmt.Println("Error cleaning cassettes:", err)
		os.Exit(1)
	}

	m.Run()

	dirty, err := snaps.Clean(m, snaps.CleanOpts{Sort: true})

	if err != nil {
		fmt.Println("Error cleaning snaps:", err)
		os.Exit(1)
	}
	if dirty {
		fmt.Println("Some snapshots were outdated.")
		os.Exit(1)
	}
}
```

A self-contained example:

```go
// main_test.go
package main

import (
	"fmt"
	"os"
	"testing"
)

func cleanSnapshots() bool {
	// (in real life this actually does stuff)
	return false
}

func TestMain(m *testing.M) {
	m.Run()

	dirty := cleanSnapshots()

	if dirty {
		fmt.Println("Some snapshots were outdated.")
		os.Exit(1)
	}
}
```

**Expected behavior**
I expect nothing to be flagged, since none of my `os.Exit` calls are based on the return from `m.Run`

**Logs**

**Desktop (please complete the following information):**
 - OS: Ubuntu 22.04 via WSLv2 Win11
 - Version of Go

**Additional context**

**Comment by ccoVeille** [comment]
I'm unsure I get the point.

For me there is no reason to use os.Exit in TestMain since Go 1.15

When they fixed this bad pattern that was breaking defer and cleanup.

> A TestMain function is no longer required to call os.Exit. If a TestMain function returns, the test binary will call os.Exit with the value returned by m.Run.
https://go.dev/doc/go1.15#testingpkgtesting

So, for me, it's normal to report the os.Exit call in TestMain as they should not be present.

But maybe revive could report them in a better way

**Comment by G-Rath** [comment]
You still need to call `os.Exit` manually in `TestMain` if you want a non-zero exit code, such as because pre-test setup or post-test cleanup failed like in my example.

The only other option would be to `panic` which would mean any top-level `defer`s get executed but has a more explosive output that could be misleading, and so I don't think it's fitting for this particular rule to enforce that (i.e if you were thinking of having the rule changed to be "use `panic` instead if you want to fail the suite"), though that could be useful to some as its own rule or an option

**Comment by G-Rath** [comment]
Also just noting that cleanup runs after the (sub)tests have completed, not when `TestMain` returns i.e. by the time `m.Run` is finished, cleanups have been run - so the use of `os.Exit` in `TestMain` only impacts `defer`s within that same `TestMain`

**Comment by ccoVeille** [comment]
You are right.

That's the only way to invalidate a test when doing a teardowm that could fail

I found these issues

https://github.com/golang/go/issues/57326

For me here, we are in a limited edge case scenario.

In your specific cases, what is reported is wrong. So it's a false-positive for you.

It could be fixed maybe, I'm unsure. It could lead to false-negative then.

Lets see what @alexandear or @chavacava think about it.

If I was you I would add a directive to ignore that line.

**Comment by G-Rath** [comment]
I think you should be able to reasonably improve the rule based on:
  - is `defer` used in `TestMain`?
  - is the return of `m.Run` assigned to a variable?

If the answer to both of those is "no", then you shouldn't need to report any `os.Exit` calls within `TestMain`, which would address this issue.

I think there's also other improvements to this rule based on different answers to those questions, such as if the answer is "yes" to the first question (regardless of the second), then any `os.Exit` calls should be reported as being an issue - but I don't think those all need to be necessarily accounted for in the one PR. 

I would also suggest renaming the rule and updating the documentation as to me the use of "redundant" implies it only cares about the pattern of `code := m.Run(); os.Exit(code)` but what we're talking about with `defer` is outside of that - so it could be more appreciate to call it something like `suspect-test-main-exit` or `risky-test-main-exit`

**Comment by alexandear** [comment]
Thanks for raising this. It's a rare case, but definitely a false positive we need to fix.

**Comment by urwashi2597** [comment]
I reproduced this on `master` (0ac11a6) and would like to work on a fix, if nobody else is on it.

The root cause looks broader than the reported case: the rule reports any exit call found in `TestMain` without checking where the exit status comes from. Since Go 1.15 the runner calls `os.Exit(m.Run())` when `TestMain` returns, so an exit is redundant only when it relays the result of `m.Run`. Three variants are currently reported although none of them is redundant:

```go
func TestMain(m *testing.M) {
	if err := setup(); err != nil {
		os.Exit(1) // reported, but this is the only way to fail the run
	}
	m.Run()
}

func TestMain(m *testing.M) {
	if err := setup(); err != nil {
		log.Fatal("setup failed") // reported
	}
	m.Run()
}

func TestMain(m *testing.M) {
	m.Run()
	syscall.Exit(3) // reported
}
```

My plan is to track the variables holding the result of `m.Run` and report an exit call only when its argument is that result, directly (`os.Exit(m.Run())`) or through such a variable (`code := m.Run(); os.Exit(code)`). Both of those stay reported.

One scope question before I open a PR: this also means restricting the check to `os.Exit`/`syscall.Exit`, the only ones taking a caller supplied status code. `log.Fatal` and friends always exit with `1`, so they can never relay the result of `m.Run` and would no longer be reported by this rule. That seems consistent with `deep-exit`, which already exempts `TestMain` from its own exit check, but it does narrow the rule, so I would rather confirm the direction first.

Happy to adjust if you would prefer to keep the fix limited strictly to `os.Exit`/`syscall.Exit` with a non-`m.Run` argument and leave `log.Fatal` reported as today.
